File: proposed-projects/round-trip-experiment/reverse/20260909T181751Z_00b43b48/20260910T095901Z_ef14bd50/ported_implementation/gbnf/rules_builder/parse_name.py
from ..utils.errors import GrammarParseError
from .is_word_char import is_word_char

PARSE_NAME_ERROR = "Failed to find a valid name"

VALID_NAME_SEPARATORS = [
    "-",
    "_",
]


def parse_name(grammar: str, pos: int) -> str:
    name = ""
    while pos < len(grammar) and (
        is_word_char(grammar[pos]) or grammar[pos] in VALID_NAME_SEPARATORS
    ):
        name += grammar[pos]
        pos += 1
    if not name:
        raise GrammarParseError(grammar, pos, PARSE_NAME_ERROR)
    return name

File: packages/gbnf-experiment/data/20260909T024849Z_b96078c6/ported_implementation/src/rules-builder/parse-name.ts
import { asCodePoints, asString } from "../utils/code-points.ts";
import { GrammarParseError } from "../utils/errors/grammar-parse-error.ts";
import { isWordChar } from "./is-word-char.ts";

export const PARSE_NAME_ERROR = "Failed to find a valid name";

export const VALID_NAME_SEPARATORS = ["-", "_"];

export const parseName = (grammar: string | string[], pos: number): string => {
  const chars = asCodePoints(grammar);
  let name = "";
  while (
    pos < chars.length &&
    (isWordChar(chars[pos]) || VALID_NAME_SEPARATORS.includes(chars[pos]))
  ) {
    name += chars[pos];
    pos += 1;
  }
  if (!name) {
    throw new GrammarParseError(asString(grammar), pos, PARSE_NAME_ERROR);
  }
  return name;
};
